keystrokes/src/components/Game/Game.js

```diff
@@ -1,98 +1,112 @@
-import React, { Component, Fragment } from 'react';
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
 import '../../stylesheets/App.css';
 
-import { getNextKey } from '../utils/utils.js';
+import {
+	getNextKey,
+	characterKeybindings,
+	validatePress,
+	verifyKey,
+} from '../utils/utils';
 
 import { ref } from '../../config/constants';
 
-class App extends Component {
-	constructor() {
-		super();
-		this.getNextKey = getNextKey.bind();
-		this.body = null;
+Game.propTypes = {
+	userInfo: PropTypes.object,
+};
+function Game({ userInfo }) {
+	const [key, setKey] = useState();
+	const [keyBindings, setKeyBindings] = useState();
+	const [failedFirstTry, setFailedFirstTry] = useState(false);
 
-		// Initial values, I think these should be set from a webpage that occurs
-		// before here. None of these db calls should be done in this class tbh
-		// they should be passed from something above (no idea how to do that yet)
-		this.state = {
-			keybindings: { gettingStarted: { spell: 'getting started!' } },
-			keys: ['gettingStarted'],
-			key: 'gettingStarted',
-		};
-	}
-
-	async componentDidMount() {
-		this.body = document.querySelector('body');
-		this.body.onkeydown = this.handleKeyPress;
+	useEffect(() => {
+		async function collectCharacterInfo() {
+			const path = `/Characters/${userInfo.selectedCharacter}`;
+			const snapShot = await ref.child(path).once('value');
+			if (snapShot.exists()) {
+				const charDetails = snapShot.val();
 
-		try {
-			const snapshot = await ref.child('/Keybindings/1').once('value');
-			const keybindings = snapshot.val();
-			//console.log('keybindings from the server', keybindings);
-
-			this.setState(
-				{
-					keybindings,
-					keys: Object.keys(keybindings),
-					key: '1',
-				},
-				() => {
-					const newKey = this.getNextKey(this.state.keys);
-					this.setState({ key: newKey });
-				}
-			);
-		} catch (error) {
-			console.warn(error);
+				collectKeybindings(
+					charDetails,
+					charDetails.selectedSpec,
+					charDetails.specs[charDetails.selectedSpec]
+						.selectedKeybindings
+				);
+			}
 		}
-	}
+		async function collectKeybindings(character, spec, keyBinding) {
+			const path = `/Keybindings/${characterKeybindings(
+				character,
+				spec,
+				keyBinding
+			)}`;
+			const snapShot = await ref.child(path).once('value');
+			if (snapShot.exists()) {
+				const k = snapShot.val();
+				setKeyBindings(k);
+				const newKey = getNextKey(Object.keys(k));
+				setKey(newKey);
+			}
+		}
+		collectCharacterInfo();
+	}, [userInfo]);
+
+	useEffect(() => {
```

Comment:
i tried to move this code into the above function, and change the function to be anonymous, but it still didnt work. I added dependancies to it so it would only run when key or keybindings would change though, so it should be fine now.